aquasurf/activation.py

```diff
@@ -0,0 +1,148 @@
+
+# Copyright (C) 2023 Cognizant Digital Business, Evolutionary AI.
+# All Rights Reserved.
+# Issued under the Academic Public License.
+#
+# You can be released from the terms, and requirements of the Academic Public
+# License by purchasing a commercial license.
+# Purchase of a commercial license is mandatory for any use of the
+# AQuaSurF Software in commercial settings.
+#
+# END COPYRIGHT
+"""
+Directed Acyclic Graph Activation Function
+"""
+import math
+import numpy as np
+from scipy import integrate
+from scipy.stats import norm
+import tensorflow as tf
+from tensorflow.keras.layers import Layer # pylint: disable=import-error, no-name-in-module
+
+N_ARY_FUNCTIONS = {
+    'sum_n' : lambda inputs : tf.math.reduce_sum(inputs, axis=0),
+    'prod_n' : lambda inputs : tf.math.reduce_prod(inputs, axis=0),
+    'max_n' : lambda inputs : tf.math.reduce_max(inputs, axis=0),
+    'min_n' : lambda inputs : tf.math.reduce_min(inputs, axis=0),
+}
+
+BINARY_FUNCTIONS = {
+    'add' : tf.math.add,
+    'sub' : tf.math.subtract,
+    'mul' : tf.math.multiply,
+    'div' : tf.math.divide,
+    'pow' : tf.math.pow,
+    'max' : tf.math.maximum,
+    'min' : tf.math.minimum,
+}
+
+UNARY_FUNCTIONS = {
+    'zero' : tf.zeros_like,
+    'one' : tf.ones_like,
+    'identity' : tf.identity,
+    'negative' : tf.math.negative,
+
+    'abs' : tf.abs,
+    'reciprocal' : tf.math.reciprocal,
+    'square' : tf.square,
+    'exp' : tf.math.exp,
+
+    'erf' : tf.math.erf,
+    'erfc' : tf.math.erfc,
+    'sinh' : tf.sinh,
+    'cosh' : tf.cosh,
+
+    'tanh' : tf.tanh,
+    'expm1' : tf.math.expm1,
+    'sigmoid' : tf.sigmoid,
+    'log_sigmoid' : tf.math.log_sigmoid,
+
+    'arcsinh' : tf.math.asinh,
+    'arctan' : tf.math.atan,
+    'bessel_i0e' : tf.math.bessel_i0e,
+    'bessel_i1e' : tf.math.bessel_i1e,
+
+    'relu' : tf.nn.relu,
+    'elu' : tf.nn.elu,
+    'selu' : tf.nn.selu,
+    'swish' : tf.nn.swish,
+
+    'softplus' : tf.nn.softplus,
+    'softsign' : tf.nn.softsign,
+    'hard_sigmoid' : tf.keras.activations.hard_sigmoid,
+}
+
+FUNCTIONS = {**N_ARY_FUNCTIONS, **BINARY_FUNCTIONS, **UNARY_FUNCTIONS}
+
+INF = float('inf')
+
+def gaussian_mean(afn):
+    return integrate.quad(lambda x : afn(x) * norm.pdf(x), -INF, INF)[0]
+
+def second_moment(afn):
+    return integrate.quad(lambda x : math.pow(afn(x), 2) * norm.pdf(x), -INF, INF)[0]
+
+class ActivationFunction(Layer):
```

Comment:
Strongly consider at least a class comment for each class.
After that consider a method comment for each public method.

Reply:
Done